### PR TITLE
Relaxed the requirement regarding atomic override of destination registers for LD to allow more implementation options

### LINKED ISSUE #51: Atomic LD writeback requirement is expensive for small cores
This extension is attractive for embedded-class cores because it significantly improves code density, and for the most part can be executed on the existing load/store unit. The only real complication is this part in section 2.1:

> The LD instruction must however write the loaded data to the pair of destination registers atomically
to ensure fault handling is possible.

Practically speaking this means that, when issuing a back-to-back pair of `lw`s, the first load can't write back to the register file until the fault response for the _second_ load comes back from the memory subsystem. Since the fault response is usually aligned with the load data coming back from the bus (i.e. is in a later pipe stage), this requires either:

* Delaying the writeback and inserting a pipe bubble after the second load (making `ld` slower than `lw` + `lw`), or
* Supporting two-register writeback, and an additional 32-bit buffer to hold the first load's data whilst waiting for the second load to complete on the bus

Neither of these is desirable for a small core with a 2R1W register file. This could be completely avoided by relaxing the constraint to something like:

> An LD instruction encountering a fault may write to at most one register in the pair `rd`, `rd + 1`. However, to make fault handling possible, an LD instruction which encounters a fault is guaranteed **not** to write back to `rs1`, even when this aliases a register in the pair `rd`, `rd + 1`.

This relaxed version can be implemented as a pair of `lw`, simply by swapping the order of the two loads based on the LSB of  register number `rs1` to ensure the first load in the pair can't clobber the base register.

### PR DESCRIPTION
**Pull request by christian-herber-nxp**
The previous definition prevented some implementations for low end MCUs, which may be desirable. Particularly, for devices with a 32b data interface and a 2R1W register file, atomic write back to the pair would only be possible by disabling interrupts during the operation. The requirement for atomic writeback was unnecessarily restrictive. This change relaxed the requirement to the minimum needed to ensure fault handling is possible. The change is backwards compatible as any implementation complying to the more restrictive requirement complies to the updated one

### CODE REVIEW

**Comment by christian-herber-nxp** [comment]
prepared this draft for review. To merged only as resolution during public review.
@Wren6991, please check if this resolution would address your issue.

**Review by tovine — APPROVED** [review]

File: zilsd.adoc
Comment (tovine) [inline zilsd.adoc]: Typo: "is" instead of "in" (the pair).

Also, this seems a bit redundant (since loads only have one source register):
> To guarantee this, if one of the destination registers of the pair is the source register containing the base, it must not be written to before the other register is the pair has been written.

Maybe something like this instead?
> To guarantee this, if one of the destination registers of the pair is also the source, it must not be written before the other register in the pair.

**Comment by Wren6991** [comment]
> prepared this draft for review. To merged only as resolution during public review. @Wren6991, please check if this resolution would address your issue.

Yes thank you, this fully addresses the issue I raised in #51. I'll leave any minor wording issues to the spec maintainers.

**Comment by GregAC** [comment]
Just wanted to add my support to this change (speaking as a maintainer of the lowRISC Ibex core: https://github.com/lowRISC/ibex)

File: zilsd.adoc
Comment (christian-herber-nxp) [reply zilsd.adoc]: I found just source a bit two ambiguous, particularly for c.ldsp, which does not have an explicit source operand. I am open to any suggestions. Meanwhile, I will fix the typo.

Comment (christian-herber-nxp) [inline zilsd.adoc]:
```suggestion
To guarantee this, if one of the destination registers of the pair is the source register containing the base, it must not be written to before the other register in the pair has been written.
```

**Comment by christian-herber-nxp** [comment]
The Arc has given the approval for this change. I would still like to ask input on the wording. Especially if a native speaker could check this, that would be great.

**Comment by Wren6991** [comment]
> I would still like to ask input on the wording.

I think the normative part is fine, and presumably this is what the ARC reviewed so we'd best not change it. I have a couple of suggestions for the informative part. Current wording:

>If an implementation performs a doubleword access atomically and the register file allows writing the pair, the mentioned atomicity requirements are inherently fulfilled.
Otherwise, an implementation either needs to delay the writeback until the write can be performed atomically,
or order sequential writes to the registers to ensure the requirement above is satisfied.

My suggested tweaks in bold:

> If an implementation performs a doubleword **load** atomically, and the register file **implements 64-bit writeback**, the mentioned atomicity requirements are inherently fulfilled.
Otherwise, **if the doubleword load is implemented as multiple bus accesses**, an implementation either needs to delay the writeback until it can be performed atomically,
or order sequential writes to the registers to ensure the requirement above is satisfied.

Reasons for suggestions:

* This discussion only applies to loads, so let's be specific (loads and stores do not necessarily have the same atomicity)
* "writing the pair" is a bit vague -- we are specifically talking about all 64 bits of load data being written to the register file in a single transaction, so I tried to spell it out a bit more
* "Otherwise" is ambiguous when preceded by two conditions; only the first is false in this clause, so specify it

**Comment by christian-herber-nxp** [comment]
> > I would still like to ask input on the wording.
> 
> I think the normative part is fine, and presumably this is what the ARC reviewed so we'd best not change it. I have a couple of suggestions for the informative part. Current wording:
> 
> > If an implementation performs a doubleword access atomically and the register file allows writing the pair, the mentioned atomicity requirements are inherently fulfilled.
> > Otherwise, an implementation either needs to delay the writeback until the write can be performed atomically,
> > or order sequential writes to the registers to ensure the requirement above is satisfied.
> 
> My suggested tweaks in bold:
> 
> > If an implementation performs a doubleword **load** atomically, and the register file **implements 64-bit writeback**, the mentioned atomicity requirements are inherently fulfilled.
> > Otherwise, **if the doubleword load is implemented as multiple bus accesses**, an implementation either needs to delay the writeback until it can be performed atomically,
> > or order sequential writes to the registers to ensure the requirement above is satisfied.
> 
> Reasons for suggestions:
> 
> * This discussion only applies to loads, so let's be specific (loads and stores do not necessarily have the same atomicity)
> * "writing the pair" is a bit vague -- we are specifically talking about all 64 bits of load data being written to the register file in a single transaction, so I tried to spell it out a bit more
> * "Otherwise" is ambiguous when preceded by two conditions; only the first is false in this clause, so specify it

I took over the first two points (374422e). On the last one, I think the text will be too convoluted if we actually make it precise. This is the else condition to the previous sentence, and it would contain another if-else. For a non-normative text, I think this is good enough, as it just provides some implementation hints.

**Comment by christian-herber-nxp** [comment]
As there is no further feedback, I would wait until tomorrow, and merge if no additional comments arise.

**Review by tovine — APPROVED** [review]

**Pull request merged**
